multistream-2.0/spec.md

```diff
@@ -0,0 +1,390 @@
+
+### Multistream Dynamic
+
+The `multistream/dynamic` protocol is used to select a protocol using a
+*receiver specified*, session-ephemeral protocol ID. These IDs are analogues of
+ephemeral ports.
+
+In this protocol, the stream initiator sends a varint ID specified by the
+*receiver* to the receiver.
+
+Format:
+
+```
+<multistream/multicodec><multistream/dynamic><id(varint)>
+```
+
+The ID 0 is reserved for saying "same protocol" on a bidirectional stream. The
+receiver of a bidirectional stream can't reuse the same dynamic ID that the
+initiator used as this dynamic ID is relative *to* the receiver. Really, this
+last rule *primarily* exists to side-step the TCP simultaneous connect issue.
+
+This protocol has *also* been designed to be hardware friendly:
+
+1. Hardware can compare the first 16 bits of the message against
+   `<multistream/multicodec><multistream/dynamic>`.
+2. It can then route the message based on the dynamic ID. The fact that these
+   IDs are chosen by the *receiver* means that the receiver can reuse the same
+   IDs for all connected peers (reusing the same hardware routing table).
+   
+### Serial Stream
+
+The `serial-stream` protocol is the simplest possible stream multiplexer.
+Unlike other stream multiplexers, `serial-stream` can only multiplex streams
+in *serial*. That is, it has to close the current stream to open a new one.
+
+The protocol is:
+
+```
+<header (signed 16 bit int)>
+<body>
+```
+
+Where the header is:
+
+* -2 - Abnormal End: Send a reset and return to multistream. All queued data (remote and
+  local) should be discarded.
+* -1 - Close: Send an EOF and return to multistream.
+*  0 - Rest: Ends the reuse protocol, transitioning to a direct stream.
+* &gt;0 - Data: The header indicates the length of the data.
+
+We could also use a varint but it's not really worth it. The 16 bit integer
+makes implementing this protocol trivial, even in hardware.
+
+Why: This allows us to:
+
+1. Try protocols and fall back on others.
+2. More importantly, it allows us to speak a bunch of protocols before setting
+   up a stream multiplexer. Specifically, we can use this for
+   `multistream/advertise` to send an advertisement as early as possible.
+
+## Upgrade Path
+
+#### Short term
+
+The short-term plan is to first negotiate multistream 1.0 and *then* negotiate
+an upgrade. That is, on connect, the *initiator* will send:
+
+```
+<len>/multistream/1.0.0\n
+<len>/multistream/2.0.0\n
+```
+
+As a batch. It will then wait for the other side to respond with either:
+
+```
+<len>/multistream/1.0.0\n
+<len>na\n
+```
+
+in which case it'll continue using multistream 1.0, or:
+
+```
+<len>/multistream/1.0.0\n
+<len>/multistream/2.0.0\n
+```
+
+in which case it'll switch to multistream 2.0.
+
+Importantly: When we switch to multistream 2.0, we'll tag the connection (and
+any sub connections) with the multistream version. This way, we never have to do
+this again.
+
+## TCP Simultaneous Open
```

Comment:
@vyzo, @jhiesey 

This is my proposal for handling simultaneous open. TL;DR: The connection has two unidirectional streams until they're joined.

@vyzo IIRC, this is slightly different from the protocol we discussed as we *always* use it, even if we have no reason to believe we performed a simultaneous connect.

Reply:
LGTM but we need more wire detail on the examples.

So let's have two examples with full protocol detail, one for the common case and one for simultaneous open.

Reply:
SGTM.